src/hotspot/share/runtime/arguments.cpp

```diff
@@ -2938,25 +2997,36 @@ jint Arguments::parse_each_vm_init_arg(const JavaVMInitArgs* args, bool* patch_m
   return JNI_OK;
 }
 
-void Arguments::add_patch_mod_prefix(const char* module_name, const char* path, bool* patch_mod_javabase) {
-  // For java.base check for duplicate --patch-module options being specified on the command line.
-  // This check is only required for java.base, all other duplicate module specifications
-  // will be checked during module system initialization.  The module system initialization
-  // will throw an ExceptionInInitializerError if this situation occurs.
-  if (strcmp(module_name, JAVA_BASE_NAME) == 0) {
-    if (*patch_mod_javabase) {
-      vm_exit_during_initialization("Cannot specify " JAVA_BASE_NAME " more than once to --patch-module");
-    } else {
-      *patch_mod_javabase = true;
-    }
-  }
+bool match_module(void *module_name, ModulePatchPath *patch) {
+  return (strcmp((char *)module_name, patch->module_name()) == 0);
+}
 
+bool Arguments::patch_mod_javabase() {
+    return _patch_mod_prefix != nullptr && _patch_mod_prefix->find((void*)JAVA_BASE_NAME, match_module) >= 0;
+}
+
+void Arguments::add_patch_mod_prefix(const char* module_name, const char* path, bool allow_append) {
   // Create GrowableArray lazily, only if --patch-module has been specified
-  if (_patch_mod_prefix == NULL) {
+  if (_patch_mod_prefix == nullptr) {
     _patch_mod_prefix = new (mtArguments) GrowableArray<ModulePatchPath*>(10, mtArguments);
   }
 
-  _patch_mod_prefix->push(new ModulePatchPath(module_name, path));
+  // Scan patches for matching module
+  int i = _patch_mod_prefix->find((void*)module_name, match_module);
+  if (i == -1) {
+    _patch_mod_prefix->push(new ModulePatchPath(module_name, path));
+  } else {
+    if (allow_append) {
+      // append path to existing module entry
+      _patch_mod_prefix->at(i)->append_path(path);
+    } else {
+      if (strcmp(module_name, JAVA_BASE_NAME) == 0) {
+        vm_exit_during_initialization("Cannot specify " JAVA_BASE_NAME " more than once to --patch-module");
+      } else {
+        vm_exit_during_initialization("Cannot specify a module more than once to --patch-module", module_name);
+      }
```

Comment:
This just reduces to:
```
else {
  vm_exit_during_initialization(err_msg("Cannot specify %s more than once to --patch-module", module_name));
}
```
or even:
```
vm_exit_during_initialization("Cannot specify a module more than once to --patch-module", module_name);
```

Reply:
Right, for some reason there are several tests that expect different error messages due to the prior implementation split between hotspot argument parsing and ModuleBootstrap.
I couldn't see clearly far enough in the EA/Preview/release cycle to be sure of the final disposition and so punted on correcting everything everywhere.
Eventually, the provision to replace some classes with value classes unnecessary. The classes will become permanent.
But that can only happen after the feature is no-longer preview and is fully integrated.